libcore/src/geometry/helper/CorrectGeometry.cpp

```diff
@@ -301,238 +301,332 @@ void CorrectInputGeometry(Building & building)
+            Point tmp{wall.GetPoint1()};
+            wall.SetPoint1(wall.GetPoint2());
+            wall.SetPoint2(tmp);
+        });
+    }
+    Point start{std::begin(trackWalls)->GetPoint1()};
+
+    for(const auto trainDoor : train._doors) {
+        double distanceFromTrackStart = trainStartOffset + trainDoor._distance;
+        double width                  = trainDoor._width;
+        std::vector<Point> intersectionPoints;
+
+        auto doorStart = FindWallPointWithDistanceOnWall(trackWalls, start, distanceFromTrackStart);
+
+        if(doorStart.has_value()) {
+            auto doorStartWallItr = std::find_if(
+                std::begin(trackWalls), std::end(trackWalls), [&doorStart](const Wall & wall) {
+                    return wall.IsInLineSegment(doorStart.value()) &&
+                           wall.GetPoint2() != doorStart.value();
+                });
+
+            std::for_each(
+                doorStartWallItr,
+                std::end(trackWalls),
+                [&width, &doorStart, &intersectionPoints](const Wall & wall) {
+                    std::vector<Point> intersections;
+                    wall.IntersectionWithCircle(doorStart.value(), width, intersections);
+                    intersectionPoints.insert(
+                        std::end(intersectionPoints),
+                        std::begin(intersections),
+                        std::end(intersections));
+                });
+
+            auto doorEndPointItr = std::find_if(
+                std::begin(intersectionPoints),
+                std::end(intersectionPoints),
+                [&trackWalls, &doorStart](const Point & intersectionPoint) {
+                    auto endItr = std::find_if(
+                        std::begin(trackWalls),
+                        std::end(trackWalls),
+                        [&intersectionPoint](const Wall & wall) {
+                            return wall.IsInLineSegment(intersectionPoint) &&
+                                   wall.GetPoint2() != intersectionPoint;
+                        });
+                    if(endItr == std::end(trackWalls)) {
+                        return false;
                     }
-                }
+                    Line tmp{doorStart.value(), endItr->GetPoint2(), 0};
+                    return tmp.IsInLineSegment(intersectionPoint);
+                });
+
+            // No intersection point in correct direction found
+            if(doorEndPointItr == std::end(intersectionPoints)) {
+                LOG_WARNING(
+                    "Compute train door coordinates of train {} at track {}: train door with "
+                    "distance {} and width {} could not be placed on track walls. Please check "
+                    "your geometry.",
+                    train._type,
+                    track._id,
+                    trainDoor._distance,
+                    trainDoor._width);
+                continue;
             }
```

Comment:
please add more comments for understanding.

Reply:
Added more comments